src/ffi/CXXInterface.h

```diff
@@ -0,0 +1,645 @@
+        Def = clang::cast<clang::ClassTemplateSpecializationDecl>(
+          Decl->getDefinition());
+      }
+      return CXXType{Def};
+    }
+
+    /**
+     * Get the template especialization with args.
+     */
+    clang::QualType getTemplateSpecializationType(
+      const clang::NamedDecl* decl,
+      llvm::ArrayRef<clang::TemplateArgument> args)
+    {
+      clang::TemplateName templ{clang::dyn_cast<clang::TemplateDecl>(
+        const_cast<clang::NamedDecl*>(decl))};
+      return ast->getTemplateSpecializationType(templ, args);
+    }
+
+    /**
+     * Get the canonical template especialization with args.
+     */
+    clang::QualType getCanonicalTemplateSpecializationType(
+      const clang::NamedDecl* decl,
+      llvm::ArrayRef<clang::TemplateArgument> args)
+    {
+      clang::TemplateName templ{clang::dyn_cast<clang::TemplateDecl>(
+        const_cast<clang::NamedDecl*>(decl))};
+      return ast->getCanonicalTemplateSpecializationType(templ, args);
+    }
+
+    /**
+     * Instantiate a new function at the end of the main file, if not yet done.
+     */
+    clang::FunctionDecl* instantiateFunction(
+      const char* name, llvm::ArrayRef<CXXType> args, CXXType ret)
+    {
+      auto* DC = ast->getTranslationUnitDecl();
+      clang::SourceLocation loc = Clang->getEndOfFileLocation();
+      clang::IdentifierInfo& fnNameIdent = ast->Idents.get(name);
+      clang::DeclarationName fnName{&fnNameIdent};
+      clang::FunctionProtoType::ExtProtoInfo EPI;
+
+      // Get type of args/ret, function
+      llvm::SmallVector<clang::QualType> argTys;
+      for (auto argTy : args)
+        argTys.push_back(getQualType(argTy));
+      auto retTy = getQualType(ret);
+      clang::QualType fnTy = ast->getFunctionType(retTy, argTys, EPI);
+
+      // Create a new function
+      auto func = clang::FunctionDecl::Create(
+        *ast,
+        DC,
+        loc,
+        loc,
+        fnName,
+        fnTy,
+        ast->getTrivialTypeSourceInfo(fnTy),
+        clang::StorageClass::SC_None);
+
+      // Associate with the translation unit
+      func->setLexicalDeclContext(DC);
+      DC->addDecl(func);
+
+      return func;
+    }
+
+    /**
+     * Create a function argument
+     *
+     * FIXME: Do we want to have this as part of instantiateFunction?
+     */
+    clang::ParmVarDecl* createFunctionArgument(
+      const char* name, CXXType& ty, clang::FunctionDecl* func)
+    {
+      clang::SourceLocation loc = func->getLocation();
+      clang::IdentifierInfo& ident = ast->Idents.get(name);
+      clang::ParmVarDecl* arg = clang::ParmVarDecl::Create(
+        *ast,
+        func,
+        loc,
+        loc,
+        &ident,
+        getQualType(ty),
+        nullptr,
+        clang::StorageClass::SC_None,
+        nullptr);
+      func->setParams({arg});
+      return arg;
+    }
+
+    /**
+     * Create integer constant literal
+     *
+     * TODO: Can we have a generic literal creator or do we need one each?
```

Comment:
I think the only place we should be using this is for template specialisations, so combining them is probably not helpful.  The caller will be post-reification in Verona and the consumer of the result needs to know that it's an IntegerLiteral.